test/Npgsql.Tests/SystemTransactionTests.cs

```diff
@@ -341,7 +489,7 @@ public void SchemaConnectionShouldntEnlist()
         }
 
         void AssertNoPreparedTransactions()
-            => Assert.That(GetNumberOfPreparedTransactions(), Is.EqualTo(0));
+            => Assert.That(GetNumberOfPreparedTransactions(), Is.EqualTo(0), "Prepared transactions found");
 
         int GetNumberOfPreparedTransactions()
```

Comment:
Yes, on NHibernate side, all our concurrency issues occurs only when the transaction is truly distributed (involving delegation to the external MsDtc process).

With NHibernate we have no "transaction commit delayed after scope disposal" issue with Sql-Server, be it through `SqlConnection` or `OdbcConnection`. But we have that with Oracle (managed driver, the others are not yet tested) and SQLite, in addition to PotsgreSql. So I think Microsoft knows what should be done but has not documented it properly, causing some other vendors to have those kind of issues. My best hope is still using a durable ressource manager instead of a volatile one. But as I have already written, I have no evidence that it is either enough or necessary.

#1580 is related to those issues: with sql-server in some circumstances re-using a connection after a transaction fails because it has not yet un-enlist from the transaction. Usually it does un-enlist by itself, but in distributed cases, sometimes it does not (or does it too late). There, enlisting with `null` allows to un-enlist from the already completed transaction and allows using the connection without a bad hack like `Thread.Sleep`.